Expose `assert_queries_match` and `assert_no_queries_match` assertions

fatkodima opened the pull request:
Follow up to https://github.com/rails/rails/pull/50281.

The originally introduced (exposed) helpers are a bit limited. For example:
1. we must specify an exact number of queries we are expecting
2. only application related queries are considered, schema/transactions related queries are always skipped
3. only one matcher can be specified

These limitations work most of the time for regular users, but not when trying to test some lower behavior (for example, inside some gem's tests).

For example, it was not possible to test that an index and a foreign key are created when running some operation.
Now, this can be done via:
```ruby
assert_queries_match(/create index/i, include_schema: true) do
  do_something
end
```

cc @byroot @p8

p8 commented on activestorage/test/models/variant_with_record_test.rb:
```diff
@@ -105,9 +105,8 @@ class ActiveStorage::VariantWithRecordTest < ActiveSupport::TestCase
-      assert_queries(6) do
-        # 6 queries:
-        # users x 1
+      assert_queries(5) do
```
Strange 🤔 Any idea why this changed?

fatkodima replied:
The `assert_queries` method originally counted for cached queries, now it is not. 
We can probably also have an option to configure if we count for cached queries or not, but not sure if it will be used much.

p8 commented:
I like that this cleans up the ActiveRecord code base and makes sure we only have one `assert_queries`.
My original PR was mostly for Rails apps. I hadn't thought of libraries using it.
Not sure if the options are too complex for a Rails app, but as these assertions are used in edge cases it might make sense to give more control.
Should the other assertions like `assert_sql` also be made public?

fatkodima commented:
> Should the other assertions like `assert_sql` also be made public?

My original idea for this PR was exactly to make `assert_sql` public, but when I investigated the current `assert_queries` implementation and realised that it can be extended, the need for `assert_sql` disappeared. I actually changed its implementation in this PR to use `assert_queries`.

byroot commented:
So in my opinion we should keep the assert relatively simple and straight forward, I'd rather avoid to cram too much in there.

We can however add a few more assertions, e.g. for the regexp we could have `assert_performed_query /SELECT ..../`.

fatkodima commented:
Wdyt if I left `assert_queries`/`assert_no_queries` as is, and introduce new `assert_sql`/`assert_no_sql` with all the needed complexity?

seanpdoyle commented on activerecord/lib/active_record/testing/query_assertions.rb:
```diff
@@ -5,29 +5,76 @@ module Assertions
+                queries
+              end
+
+            if expected_count.nil?
+              assert_operator matched_queries.size, :>=, 1, "1 or more queries expected, but none were executed."
+            else
+              assert_equal expected_count, matched_queries.size, "#{matched_queries.size} instead of #{expected_count} queries were executed. Queries: #{matched_queries.join("\n\n")}"
+            end
+          end
```
Would it be possible to support `Range` instances too?

Something like `assert_queries(2..3) { ... }`. It'd also support endless Ranges: `assert_queries(..3) { ... }`.

```suggestion
            case expected_count
            when NilClass
              assert_operator matched_queries.size, :>=, 1, "1 or more queries expected, but none were executed."
            when Range
              assert_includes expected_count, matched_queries.size, "expected #{matched_queries.size} executed queries to be within #{expected_count}. Queries: #{matched_queries.join("\n\n")}"
            else
              assert_equal expected_count, matched_queries.size, "#{matched_queries.size} instead of #{expected_count} queries were executed. Queries: #{matched_queries.join("\n\n")}"
            end
          end
```

fatkodima replied:
Can you provide a use case for this?

seanpdoyle replied:
The main motivation is flexibility and maintenance.

I'm imagining introducing this assertion into an existing application test suite in the hopes that it'd drive efforts to minimize queries (either as an optimization or as a reaction to issues like N+1s). In this hypothetical application, I might introduce the assertion with a specific query, pass CI, then merge my change. A teammate might make a change that would make a query and the test might fail. They could update the assertion to use the more correct number, pass CI, then merge their change.

Depending on the goals of the team using this assertion that type of churn might start a virtuous cycle that forces the team to change the test's number for necessary queries. Alternatively, it might change often enough to develop a negative perception, feel onerous, and ultimately get removed.

In cases where the assertion is used as an optimization, teams might be interested in providing a `Range` (like "fewer than 10" or `..10`) that they could use as a arbitrary goal.

Personally, my preference is for the specific, exact count version. In my experiences, constant churn can deter that style of usage.

fatkodima replied:
> Personally, my preference is for the specific, exact count version.

This is my preference too. Let's see what other people say.

byroot commented:
> Wdyt if I left `assert_queries`/`assert_no_queries` as is, and introduce new `assert_sql`/`assert_no_sql` with all the needed complexity?

Yeah, I'd prefer that. As long as `assert_sql` doesn't do too much. I think taking a single regexp would be plenty. If you need to match a second one, you can nest them. If you need more than that, you can write a custom assertion IMO.

byroot commented:
Just to clarify my stance, I think we should aim for the 20/80 here. 20% of the features that are sufficient for 80% of users.

fatkodima commented:
Made some changes in a separate commit - reverted an ability to provide multiple regexes. Currently (as before), it allows only one regexp.

Now, the only changes this PR introduces is the ability to not pass an exact number of queries and ability to account for internal queries (like `TRANSACTION`, schema related queries). 

Considering, that these changes are now pretty simple imo and even cover everything from `assert_sql` (except a nicer message for multiple regexes), I wonder if we need any other assertion methods? 🤔 

If this is ok, I will squash my commits.

byroot commented on activerecord/lib/active_record/testing/query_assertions.rb:
```diff
@@ -5,29 +5,65 @@ module Assertions
+      #
+      #   assert_queries(ignore_none: true) { Post.columns }
+      #
+      def assert_queries(expected_count = nil, matcher: nil, ignore_none: false, &block)
```
I really don't like `ignore_none` as a parameter name, as it's a double negation.

Also I wonder that the use case is?

byroot commented:
I still think it would be much simpler to use if that capability was in another `assert_` method.

fatkodima commented:
>  if that capability was in another `assert_` method.

Can you explain, which exact capability do you mean?

Yeah, I was thinking about `assert_sql` as another assertion, but that leads to another confusion, imo, as what is the difference between `assert_queries` and `assert_sql` then, considering that both are queries and that `assert_queries` already support `matcher` argument.

I would appreciate, if you will help me separate capabilities between `assert_queries` and some other assertion method (and give it a name) 🙏

fatkodima commented:
Updated `assert_sql` to simply check for the number of queries and added `assert_sql` to be able to do more sophisticated checks, like for regexes, internal queries etc. Please, take a look.

seanpdoyle commented on activerecord/lib/active_record/testing/query_assertions.rb:
```diff
@@ -7,28 +7,102 @@ module QueryAssertions
+      #
+      #   assert_sql(include_all: true) { Post.columns }
+      #
+      def assert_sql(expected_count = nil, matcher: nil, include_all: false, &block)
```
How do you feel about renaming the `:matcher` keyword argument to `:match`? It'd mirror the new `:match` keyword argument that [assert_raises](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Testing/Assertions.html#method-i-assert_raises) supports, and would correspond to the [Regexp#match](https://ruby-doc.org/core-3.0.1/Regexp.html#method-i-match) method name.

fatkodima replied:
👍

p8 commented:
If `assert_sql` requires a match argument we could rename the assertion and make the Regexp a regular required comment:
```ruby
assert_sql_match(/create index/i, include_all: true) do
  do_something
end
```

Or make it more inline with `assert_queries`:
```ruby
assert_query_match(/create index/i, include_all: true) do
  do_something
end
```

fatkodima commented:
> If `assert_sql` requires a match argument
 
Currently, it does not require a `matcher` argument.

byroot commented on activerecord/lib/active_record/testing/query_assertions.rb:
```diff
@@ -7,28 +7,102 @@ module QueryAssertions
+      #
+      # If the +:matcher+ option is provided, only queries that match the matcher are counted.
+      #
+      #   assert_no_sql(1, match: /LIMIT \?/) { Post.first }
```
You probably meant:

```suggestion
      #   assert_no_sql(match: /LIMIT \?/) { Post.first }
```

byroot commented on activerecord/lib/active_record/testing/query_assertions.rb:
```diff
@@ -7,28 +7,102 @@ module QueryAssertions
+      #
+      #   assert_sql(include_all: true) { Post.columns }
+      #
+      def assert_sql(expected_count = nil, match: nil, include_all: false, &block)
```
I think I'd inverse the parameters:

```suggestion
      def assert_sql(match, count: nil, include_all: false, &block)
```

What do you think?

fatkodima replied:
I originally thought for `match` to not be required to be able to write something like `assert_sql(include_all: true)` to test that *any* SQL query was made.

byroot replied:
IMO `assert_sql` means I want to assert some query matches something. 

If I only care about the count, it should be `assert_queries`.

fatkodima replied:
Yeah, I agree with you. There is a need to update quite a few test cases to specify these concrete sqls, like in https://github.com/rails/rails/pull/50373/files#diff-9bc4b7719b110cde3044f7e2ce6f8cda8f8e3263790d600e6d6792e40c02fa70R831

Will try to update with the suggestion.

p8 replied:
Also, does the count apply to the matching queries or all queries?

fatkodima replied:
Made matcher a required argument.

fatkodima committed:
Expose `assert_queries_match` and `assert_no_queries_match` assertions

byroot commented on activerecord/test/cases/migration_test.rb:
```diff
@@ -1437,7 +1437,7 @@ def test_changing_column_null_with_default
       }
 
-      assert_queries(expected_query_count, ignore_none: true) do
+      assert_sql(//, count: expected_query_count, include_all: true) do
```
This suggest `assert_queries` should have an `include_all` too.

byroot commented:
Starting to look good to me. Only the `assert_sql(//` use strike me as not ideal, all the rest seem sensible.

fatkodima commented:
Added `include_all` to `assert_queries`. Also made query count optional to be able to easier write something like https://github.com/rails/rails/pull/50373/files#diff-9bc4b7719b110cde3044f7e2ce6f8cda8f8e3263790d600e6d6792e40c02fa70R831 where we do not need exact counts and they are even different for each adapter. Let me know what you think about this.

p8 commented:
If I only see `assert_sql` and `assert_queries` it's unclear to me which one accepts a matcher.
Also I would need to look up `include_all` to see what is normally excluded.
Not sure if two seperate options`:include_transactions` and `:include_schema` is too much, but most of the time you'd probably only need to set one to true.

byroot commented:
> `:include_transactions` and `:include_schema` is too much

I'm not even sure ignoring transactions make that much sense. Looking at our private implementation, it ignore schema queries by default (because they tend to be flaky), but transactions are included.

And IMO it makes sense because `begin; query_1; query_2; commit` isn't the same thing as just `query_1; query_2`.

fatkodima commented:
1.
> If I only see `assert_sql` and `assert_queries` it's unclear to me which one accepts a matcher.

What if then we rename `assert_sql` to `assert_queries_match`? Or maybe there is a better name?

2. If we are going to change that now transactions queries will not change, should we rename `:include_all` to something else? `:include_schema` etc?

byroot commented:
> What if then we rename `assert_sql` to `assert_queries_match`? Or maybe there is a better name?

Or `assert_queries` to `assert_queries_count`. Then `assert_sql` can become `assert_query_match`?

Just a though, no strong opinion here.

> If we are going to change that now transactions queries are accounted by default, should we rename :include_all to something else? :include_schema etc?

Yeah that's the idea.

p8 commented:
I like `assert_queries_count` and `assert_query_match`.

fatkodima commented:
> assert_query_match

You probably meant `assert_queries_match`? On par with `assert_queries_count`

fatkodima commented:
Changed names:
`assert_queries` -> `assert_queries_count`
`assert_sql` -> `assert_queries_match`
`assert_no_sql` -> `assert_no_queries_match`
`include_all` -> `include_schema`

and transaction related queries are now always counted.

byroot commented:
LGTM. What do you think @p8 ?

p8 commented:
Look great!

fatkodima commented:
Thank you all! ❤️ That was quite a lengthy discussion 💪 😅

p8 commented:
Thanks for the work @fatkodima ! 😄
